File: js/src/browser/browser-cookies.js
import path from 'node:path';
import os from 'node:os';
import { TextDecoder } from 'node:util';

import Database from 'better-sqlite3';

import {
  clearBrowserCookieMemoryCache,
  getCachedCredential,
  normalizeCookieCache,
  readCookieResultCache,
  writeCookieResultCache,
} from './browser-cookie-cache.js';
import {
  chromiumSameSite,
  decryptChromiumCookie,
  deriveChromiumCookieKey,
  firefoxSameSite,
} from './browser-cookie-crypto.js';
import {
  decryptWindowsDpapi,
  readSafeStoragePassword,
  readWindowsEncryptionKey,
} from './browser-cookie-credentials.js';
import {
  findCookieDatabase,
  listBrowserProfiles,
  normalizeCookieBrowser,
  resolveBrowserProfile,
} from './browser-profiles.js';

const CHROME_EPOCH_OFFSET_SECONDS = 11_644_473_600n;
const MICROSECONDS_PER_SECOND = 1_000_000n;

function toNumber(value) {
  return typeof value === 'bigint' ? Number(value) : Number(value ?? 0);
}

function chromiumExpires(value) {
  const microseconds = BigInt(value ?? 0);
  if (microseconds === 0n) {
    return -1;
  }
  return Number(
    microseconds / MICROSECONDS_PER_SECOND - CHROME_EPOCH_OFFSET_SECONDS
  );
}

function firefoxExpires(value) {
  const expires = toNumber(value);
  return expires > 0 ? expires : -1;
}

function queryRows(database, query, domainFilter) {
  const statement = database.prepare(query).safeIntegers();
  return domainFilter ? statement.all(`%${domainFilter}%`) : statement.all();
}

function readDatabaseVersion(database) {
  try {
    const row = database
      .prepare("SELECT value FROM meta WHERE key = 'version'")
      .get();
    return Number(row?.value ?? 0);
  } catch {
    return 0;
  }
}

function readFirefoxRows(database, domainFilter) {
  return queryRows(
    database,
    `SELECT name, value, host, path, expiry, isSecure, isHttpOnly, sameSite
       FROM moz_cookies
      ${domainFilter ? 'WHERE host LIKE ?' : ''}
      ORDER BY host, name, path`,
    domainFilter
  );
}

function readChromiumRows(database, domainFilter) {
  return queryRows(
    database,
    `SELECT host_key, name, value, encrypted_value, path, expires_utc,
            is_secure, is_httponly, samesite
       FROM cookies
      ${domainFilter ? 'WHERE host_key LIKE ?' : ''}
      ORDER BY host_key, name, path`,
    domainFilter
  );
}

function mapFirefoxRows(rows) {
  return rows.map((row) => ({
    name: row.name,
    value: row.value,
    domain: row.host,
    path: row.path || '/',
    expires: firefoxExpires(row.expiry),
    httpOnly: Boolean(row.isHttpOnly),
    secure: Boolean(row.isSecure),
    sameSite: firefoxSameSite(row.sameSite),
  }));
}

function encryptionPrefix(encryptedValue) {
  return Buffer.from(encryptedValue).subarray(0, 3).toString('ascii');
}

function chromiumKeyForPrefix(context, prefix) {
  if (context.platform === 'linux' && prefix === 'v10') {
    return deriveChromiumCookieKey('peanuts', 'linux');
  }
  if (context.platform === 'linux' || context.platform === 'darwin') {
    return getCachedCredential({
      cache: context.cache,
      identity: `${context.browser}:${context.platform}:safe-storage`,
      refresh: context.refresh,
      now: context.now,
      metadata: {
        browser: context.browser,
        platform: context.platform,
        source: 'safe-storage',
      },
      create: async () =>
        deriveChromiumCookieKey(
          await context.readSafeStoragePassword({
            browser: context.browser,
            platform: context.platform,
            environment: context.environment,
          }),
          context.platform
        ),
    });
  }
  if (context.platform === 'win32') {
    return getCachedCredential({
      cache: context.cache,
      identity: `${context.browser}:win32:legacy-aes-key`,
      refresh: context.refresh,
      now: context.now,
      metadata: {
        browser: context.browser,
        platform: context.platform,
        source: 'dpapi',
      },
      create: () =>
        context.readWindowsEncryptionKey({
          localStatePath: path.join(
            path.dirname(context.profile.path),
            'Local State'
          ),
          environment: context.environment,
          decryptDpapi: context.decryptWindowsDpapi,
        }),
    });
  }
  throw new Error(
    `Chromium cookie decryption is unsupported on ${context.platform}`
  );
}

async function decryptChromiumRow(row, databaseVersion, context) {
  if (row.value) {
    return row.value;
  }
  const encryptedValue = Buffer.from(row.encrypted_value);
  if (encryptedValue.length === 0) {
    return '';
  }
  const prefix = encryptionPrefix(encryptedValue);
  if (context.platform === 'win32' && prefix !== 'v10' && prefix !== 'v11') {
    if (prefix === 'v20') {
      return decryptChromiumCookie({
        encryptedValue,
        host: row.host_key,
        databaseVersion,
        platform: context.platform,
        key: Buffer.alloc(32),
      });
    }
    const plaintext = await context.decryptWindowsDpapi(encryptedValue, {
      environment: context.environment,
    });
    return new TextDecoder('utf8', { fatal: true }).decode(plaintext);
  }
  const key = await chromiumKeyForPrefix(context, prefix);
  return decryptChromiumCookie({
    encryptedValue,
    host: row.host_key,
    databaseVersion,
    platform: context.platform,
    key,
  });
}

async function mapChromiumRows(rows, databaseVersion, context) {
  const cookies = [];
  for (const row of rows) {
    try {
      cookies.push({
        name: row.name,
        value: await decryptChromiumRow(row, databaseVersion, context),
        domain: row.host_key,
        path: row.path || '/',
        expires: chromiumExpires(row.expires_utc),
        httpOnly: Boolean(row.is_httponly),
        secure: Boolean(row.is_secure),
        sameSite: chromiumSameSite(row.samesite),
      });
    } catch (error) {
      if (!context.ignoreDecryptionErrors) {
        throw new Error(
          `Could not decrypt cookie ${row.name} for ${row.host_key}: ${error.message}`
        );
      }
    }
  }
  return cookies;
}

function openCookieDatabase(cookiePath) {
  try {
    return new Database(cookiePath, { fileMustExist: true, readonly: true });
  } catch (error) {
    throw new Error(`Could not open browser cookie database: ${error.message}`);
  }
}

/**
 * Read cookies from an installed browser profile in Playwright/Puppeteer shape.
 * Use dependency injection only for deterministic platform and credential tests.
 */
export async function readBrowserCookiesWithDependencies(
  options,
  dependencies = {}
) {
  if (!options || typeof options !== 'object') {
    throw new TypeError('readBrowserCookies requires an options object');
  }
  const browser = normalizeCookieBrowser(options.browser);
  const platform = dependencies.platform ?? process.platform;
  const homeDir = dependencies.homeDir ?? os.homedir();
  const environment = dependencies.environment ?? process.env;
  const profile = await resolveBrowserProfile({
    browser,
    profile: options.profile,
    platform,
    homeDir,
    environment,
  });
  const cookiePath = await findCookieDatabase(browser, profile.path, platform);
  if (!cookiePath) {
    throw new Error(`No cookie database exists in ${profile.path}`);
  }
  const cache = normalizeCookieCache(
    options.cache,
    homeDir,
    options.ttlMinutes
  );
  const identity = JSON.stringify({
    browser,
    profile: profile.path,
    domainFilter: options.domainFilter ?? null,
  });
  const now = dependencies.now ?? Date.now;
  const cachedCookies = await readCookieResultCache({
    cache,
    identity,
    refresh: options.refresh === true,
    now,
  });
  if (cachedCookies) {
    return cachedCookies;
  }

  const database = openCookieDatabase(cookiePath);
  let cookies;
  try {
    if (browser === 'firefox') {
      cookies = mapFirefoxRows(readFirefoxRows(database, options.domainFilter));
    } else {
      const databaseVersion = readDatabaseVersion(database);
      const rows = readChromiumRows(database, options.domainFilter);
      cookies = await mapChromiumRows(rows, databaseVersion, {
        browser,
        cache,
        decryptWindowsDpapi:
          dependencies.decryptWindowsDpapi ?? decryptWindowsDpapi,
        environment,
        ignoreDecryptionErrors: options.ignoreDecryptionErrors === true,
        now,
        platform,
        profile,
        readSafeStoragePassword:
          dependencies.readSafeStoragePassword ?? readSafeStoragePassword,
        readWindowsEncryptionKey:
          dependencies.readWindowsEncryptionKey ?? readWindowsEncryptionKey,
        refresh: options.refresh === true,
      });
    }
  } finally {
    database.close();
  }
  await writeCookieResultCache({ cache, identity, cookies, now });
  return cookies;
}

/** Read cookies from an installed Chrome, Edge, Brave, Chromium, or Firefox. */
export function readBrowserCookies(options) {
  return readBrowserCookiesWithDependencies(options);
}

export {
  clearBrowserCookieMemoryCache,
  decryptChromiumCookie,
  listBrowserProfiles,
};

File: rust/src/browser/browser_cookies.rs
//! Import cookies from installed Chrome-family and Firefox profiles.

use std::collections::HashMap;
use std::path::PathBuf;

use anyhow::{anyhow, Context, Result};
use rusqlite::{params, Connection, OpenFlags};
use serde::{Deserialize, Serialize};
use serde_json::{json, Map, Value};

use super::browser_cookie_cache::{
    get_cached_credential, normalize_cookie_cache, read_cookie_result_cache,
    write_cookie_result_cache, NormalizedCookieCache,
};
use super::browser_cookie_credentials::{
    decrypt_windows_dpapi, read_safe_storage_password, read_windows_encryption_key,
};
use super::browser_cookie_crypto::{
    chromium_same_site, decode_chromium_plaintext, decrypt_chromium_cookie,
    derive_chromium_cookie_key, firefox_same_site,
};
use super::browser_profiles::{
    find_cookie_database, normalize_cookie_browser, resolve_browser_profile, BrowserProfile,
    BrowserProfileOptions,
};

const CHROME_EPOCH_OFFSET_SECONDS: i64 = 11_644_473_600;

/// A browser cookie in the shape accepted by Playwright/Puppeteer contexts.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BrowserCookie {
    /// Cookie name.
    pub name: String,
    /// Decrypted cookie value.
    pub value: String,
    /// Cookie domain, including any leading dot.
    pub domain: String,
    /// Cookie path.
    pub path: String,
    /// Unix expiry seconds, or `-1` for a session cookie.
    pub expires: i64,
    /// Whether JavaScript is prevented from reading the cookie.
    pub http_only: bool,
    /// Whether the cookie is restricted to secure transports.
    pub secure: bool,
    /// `Strict`, `Lax`, or `None`.
    pub same_site: String,
}

/// Options for [`read_browser_cookies`].
#[derive(Debug, Clone)]
pub struct BrowserCookieReadOptions {
    /// Installed browser name.
    pub browser: String,
    /// Optional on-disk or display profile name.
    pub profile: Option<String>,
    /// Optional domain substring used by the SQLite query.
    pub domain_filter: Option<String>,
    /// Enable the owner-only decrypted-result and derived-key cache.
    pub cache: bool,
    /// Override the cache directory.
    pub cache_dir: Option<PathBuf>,
    /// Cache lifetime in minutes.
    pub ttl_minutes: Option<f64>,
    /// Bypass cached values and coordinate one refreshed credential read.
    pub refresh: bool,
    /// Skip individual cookies that cannot be decrypted.
    pub ignore_decryption_errors: bool,
    /// Home directory used for profile discovery and the default cache.
    pub home_dir: PathBuf,
    /// Platform convention (`linux`, `darwin`, or `win32`).
    pub platform: String,
}

impl BrowserCookieReadOptions {
    /// Create options for one installed browser.
    pub fn new(browser: impl Into<String>) -> Self {
        Self {
            browser: browser.into(),
            profile: None,
            domain_filter: None,
            cache: true,
            cache_dir: None,
            ttl_minutes: None,
            refresh: false,
            ignore_decryption_errors: false,
            home_dir: dirs::home_dir().unwrap_or_else(|| PathBuf::from(".")),
            platform: super::browser_profiles::current_platform().to_string(),
        }
    }

    /// Select a named installed-browser profile.
    pub fn profile(mut self, profile: impl Into<String>) -> Self {
        self.profile = Some(profile.into());
        self
    }

    /// Restrict the SQLite query to domains containing this value.
    pub fn domain_filter(mut self, domain: impl Into<String>) -> Self {
        self.domain_filter = Some(domain.into());
        self
    }

    /// Enable or disable disk caching.
    pub fn cache(mut self, enabled: bool) -> Self {
        self.cache = enabled;
        self
    }

    /// Override the owner-only cache directory.
    pub fn cache_dir(mut self, directory: impl Into<PathBuf>) -> Self {
        self.cache_dir = Some(directory.into());
        self
    }

    /// Set the decrypted-result and derived-key cache TTL.
    pub fn ttl_minutes(mut self, minutes: f64) -> Self {
        self.ttl_minutes = Some(minutes);
        self
    }

    /// Force a coordinated refresh of cached results and credentials.
    pub fn refresh(mut self, refresh: bool) -> Self {
        self.refresh = refresh;
        self
    }

    /// Skip cookies whose platform decryption fails.
    pub fn ignore_decryption_errors(mut self, ignore: bool) -> Self {
        self.ignore_decryption_errors = ignore;
        self
    }

    /// Override the home directory used for discovery.
    pub fn home_dir(mut self, home_dir: impl Into<PathBuf>) -> Self {
        self.home_dir = home_dir.into();
        self
    }

    /// Override the platform convention, primarily for portable tooling/tests.
    pub fn platform(mut self, platform: impl AsRef<str>) -> Self {
        self.platform = super::browser_profiles::normalize_platform(platform.as_ref()).to_string();
        self
    }
}

#[derive(Debug)]
struct ChromiumRow {
    host: String,
    name: String,
    value: String,
    encrypted_value: Vec<u8>,
    path: String,
    expires: i64,
    secure: bool,
    http_only: bool,
    same_site: i64,
}

#[derive(Default)]
struct OperationKeyCache {
    attempts: HashMap<String, std::result::Result<Vec<u8>, String>>,
}

struct CookieDecryptionState<'a> {
    cache: &'a NormalizedCookieCache,
    operation_keys: OperationKeyCache,
}

impl OperationKeyCache {
    fn get_or_try_create<F>(&mut self, identity: &str, create: F) -> Result<Vec<u8>>
    where
        F: FnOnce() -> Result<Vec<u8>>,
    {
        if !self.attempts.contains_key(identity) {
            self.attempts.insert(
                identity.to_string(),
                create().map_err(|error| error.to_string()),
            );
        }
        self.attempts[identity]
            .clone()
            .map_err(|error| anyhow!(error))
    }
}

fn open_cookie_database(path: &std::path::Path) -> Result<Connection> {
    Connection::open_with_flags(path, OpenFlags::SQLITE_OPEN_READ_ONLY)
        .with_context(|| format!("Could not open browser cookie database: {}", path.display()))
}

fn read_database_version(database: &Connection) -> i64 {
    database
        .query_row("SELECT value FROM meta WHERE key = 'version'", [], |row| {
            row.get::<_, String>(0)
        })
        .ok()
        .and_then(|value| value.parse().ok())
        .unwrap_or_default()
}

fn domain_pattern(domain_filter: Option<&str>) -> Option<String> {
    domain_filter.map(|domain| format!("%{domain}%"))
}

fn read_chromium_rows(
    database: &Connection,
    domain_filter: Option<&str>,
) -> Result<Vec<ChromiumRow>> {
    let where_clause = domain_filter
        .map(|_| " WHERE host_key LIKE ?1")
        .unwrap_or("");
    let query = format!(
        "SELECT host_key, name, value, encrypted_value, path, expires_utc, \
         is_secure, is_httponly, samesite FROM cookies{where_clause} \
         ORDER BY host_key, name, path"
    );
    let mut statement = database.prepare(&query)?;
    let pattern = domain_pattern(domain_filter);
    let mapper = |row: &rusqlite::Row<'_>| {
        Ok(ChromiumRow {
            host: row.get(0)?,
            name: row.get(1)?,
            value: row.get(2)?,
            encrypted_value: row.get(3)?,
            path: row.get(4)?,
            expires: row.get(5)?,
            secure: row.get::<_, i64>(6)? != 0,
            http_only: row.get::<_, i64>(7)? != 0,
            same_site: row.get(8)?,
        })
    };
    let rows = if let Some(pattern) = pattern.as_deref() {
        statement.query_map(params![pattern], mapper)?
    } else {
        statement.query_map([], mapper)?
    };
    rows.collect::<rusqlite::Result<Vec<_>>>()
        .map_err(Into::into)
}

fn read_firefox_cookies(
    database: &Connection,
    domain_filter: Option<&str>,
) -> Result<Vec<BrowserCookie>> {
    let where_clause = domain_filter.map(|_| " WHERE host LIKE ?1").unwrap_or("");
    let query = format!(
        "SELECT name, value, host, path, expiry, isSecure, isHttpOnly, sameSite \
         FROM moz_cookies{where_clause} ORDER BY host, name, path"
    );
    let mut statement = database.prepare(&query)?;
    let pattern = domain_pattern(domain_filter);
    let mapper = |row: &rusqlite::Row<'_>| {
        let expires = row.get::<_, i64>(4)?;
        Ok(BrowserCookie {
            name: row.get(0)?,
            value: row.get(1)?,
            domain: row.get(2)?,
            path: row.get::<_, String>(3).map(|path| {
                if path.is_empty() {
                    "/".to_string()
                } else {
                    path
                }
            })?,
            expires: if expires > 0 { expires } else { -1 },
            secure: row.get::<_, i64>(5)? != 0,
            http_only: row.get::<_, i64>(6)? != 0,
            same_site: firefox_same_site(row.get(7)?).to_string(),
        })
    };
    let rows = if let Some(pattern) = pattern.as_deref() {
        statement.query_map(params![pattern], mapper)?
    } else {
        statement.query_map([], mapper)?
    };
    rows.collect::<rusqlite::Result<Vec<_>>>()
        .map_err(Into::into)
}

fn credential_metadata(browser: &str, platform: &str, source: &str) -> Map<String, Value> {
    let mut metadata = Map::new();
    metadata.insert("browser".into(), json!(browser));
    metadata.insert("platform".into(), json!(platform));
    metadata.insert("source".into(), json!(source));
    metadata
}

fn chromium_key_for_prefix(
    prefix: &[u8],
    browser: &str,
    platform: &str,
    profile: &BrowserProfile,
    refresh: bool,
    state: &mut CookieDecryptionState<'_>,
) -> Result<Vec<u8>> {
    if platform == "linux" && prefix == b"v10" {
        return derive_chromium_cookie_key("peanuts", "linux");
    }
    if platform == "linux" || platform == "darwin" {
        let identity = format!("{browser}:{platform}:safe-storage");
        return state.operation_keys.get_or_try_create(&identity, || {
            get_cached_credential(
                state.cache,
                &identity,
                refresh,
                credential_metadata(browser, platform, "safe-storage"),
                || {
                    derive_chromium_cookie_key(
                        &read_safe_storage_password(browser, platform)?,
                        platform,
                    )
                },
            )
        });
    }
    if platform == "win32" {
        let identity = format!("{browser}:win32:legacy-aes-key");
        return state.operation_keys.get_or_try_create(&identity, || {
            get_cached_credential(
                state.cache,
                &identity,
                refresh,
                credential_metadata(browser, platform, "dpapi"),
                || {
                    read_windows_encryption_key(
                        &profile
                            .path
                            .parent()
                            .unwrap_or(&profile.path)
                            .join("Local State"),
                    )
                },
            )
        });
    }
    Err(anyhow!(
        "Chromium cookie decryption is unsupported on {platform}"
    ))
}

fn chromium_expires(value: i64) -> i64 {
    if value == 0 {
        -1
    } else {
        value / 1_000_000 - CHROME_EPOCH_OFFSET_SECONDS
    }
}

fn decrypt_chromium_row(
    row: ChromiumRow,
    database_version: i64,
    browser: &str,
    platform: &str,
    profile: &BrowserProfile,
    refresh: bool,
    state: &mut CookieDecryptionState<'_>,
) -> Result<BrowserCookie> {
    let value = if !row.value.is_empty() {
        row.value
    } else if row.encrypted_value.is_empty() {
        String::new()
    } else {
        let prefix = row.encrypted_value.get(..3).unwrap_or_default();
        if platform == "win32" && prefix != b"v10" && prefix != b"v11" {
            if prefix == b"v20" {
                decrypt_chromium_cookie(
                    &row.encrypted_value,
                    &row.host,
                    database_version,
                    platform,
                    &[0_u8; 32],
                )?
            } else {
                decode_chromium_plaintext(
                    &decrypt_windows_dpapi(&row.encrypted_value)?,
                    &row.host,
                    database_version,
                )?
            }
        } else {
            let key = chromium_key_for_prefix(prefix, browser, platform, profile, refresh, state)?;
            decrypt_chromium_cookie(
                &row.encrypted_value,
                &row.host,
                database_version,
                platform,
                &key,
            )?
        }
    };
    Ok(BrowserCookie {
        name: row.name,
        value,
        domain: row.host,
        path: if row.path.is_empty() {
            "/".into()
        } else {
            row.path
        },
        expires: chromium_expires(row.expires),
        http_only: row.http_only,
        secure: row.secure,
        same_site: chromium_same_site(row.same_site).to_string(),
    })
}

fn read_chromium_cookies(
    database: &Connection,
    profile: &BrowserProfile,
    options: &BrowserCookieReadOptions,
    cache: &NormalizedCookieCache,
) -> Result<Vec<BrowserCookie>> {
    let version = read_database_version(database);
    let mut cookies = Vec::new();
    let mut state = CookieDecryptionState {
        cache,
        operation_keys: OperationKeyCache::default(),
    };
    for row in read_chromium_rows(database, options.domain_filter.as_deref())? {
        let name = row.name.clone();
        let host = row.host.clone();
        match decrypt_chromium_row(
            row,
            version,
            &options.browser,
            &options.platform,
            profile,
            options.refresh,
            &mut state,
        ) {
            Ok(cookie) => cookies.push(cookie),
            Err(_) if options.ignore_decryption_errors => {}
            Err(error) => {
                return Err(anyhow!(
                    "Could not decrypt cookie {name} for {host}: {error}"
                ))
            }
        }
    }
    Ok(cookies)
}

/// Read cookies from an installed Chrome, Edge, Brave, Chromium, or Firefox profile.
pub fn read_browser_cookies(mut options: BrowserCookieReadOptions) -> Result<Vec<BrowserCookie>> {
    options.browser = normalize_cookie_browser(&options.browser)?.to_string();
    let discovery = BrowserProfileOptions::default()
        .browser(&options.browser)
        .home_dir(&options.home_dir)
        .platform(&options.platform);
    let profile =
        resolve_browser_profile(&options.browser, options.profile.as_deref(), &discovery)?;
    let cookie_path = find_cookie_database(&options.browser, &profile.path)
        .ok_or_else(|| anyhow!("No cookie database exists in {}", profile.path.display()))?;
    let cache = normalize_cookie_cache(
        options.cache,
        options.cache_dir.as_deref(),
        &options.home_dir,
        options.ttl_minutes,
    )?;
    let identity = serde_json::to_string(&json!({
        "browser": options.browser,
        "profile": profile.path,
        "domainFilter": options.domain_filter,
    }))?;
    if let Some(values) = read_cookie_result_cache(&cache, &identity, options.refresh) {
        return values
            .into_iter()
            .map(serde_json::from_value)
            .collect::<serde_json::Result<Vec<_>>>()
            .context("cached cookies have an invalid shape");
    }

    let database = open_cookie_database(&cookie_path)?;
    let cookies = if options.browser == "firefox" {
        read_firefox_cookies(&database, options.domain_filter.as_deref())?
    } else {
        read_chromium_cookies(&database, &profile, &options, &cache)?
    };
    let serialized = cookies
        .iter()
        .map(serde_json::to_value)
        .collect::<serde_json::Result<Vec<_>>>()?;
    write_cookie_result_cache(&cache, &identity, &serialized)?;
    Ok(cookies)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn operation_key_cache_reads_a_refreshed_credential_once() -> Result<()> {
        let mut keys = OperationKeyCache::default();
        let mut calls = 0;
        assert_eq!(
            keys.get_or_try_create("safe-storage", || {
                calls += 1;
                Ok(vec![7_u8; 16])
            })?,
            vec![7_u8; 16]
        );
        assert_eq!(
            keys.get_or_try_create("safe-storage", || {
                calls += 1;
                Ok(vec![8_u8; 16])
            })?,
            vec![7_u8; 16]
        );
        assert_eq!(calls, 1);
        Ok(())
    }
}
